packages/toolpad-app/src/server/data.ts

```diff
@@ -455,3 +461,43 @@ export async function loadDom(appId: string, version: VersionOrPreview = 'previe
 export async function loadRenderTree(appId: string, version: VersionOrPreview = 'preview') {
   return appDom.createRenderTree(await loadDom(appId, version));
 }
+
+const LATEST_RELEASE_API_URL = 'https://api.github.com/repos/mui/mui-toolpad/releases/latest';
+
+const latestReleaseCache: GithubReleaseCache = {
+  nextFetchAllowedAt: Number.NEGATIVE_INFINITY,
+  release: null,
+};
+
+export async function getLatestToolpadRelease(): Promise<GithubRelease> {
+  const timestamp = Date.now();
+  if (latestReleaseCache.nextFetchAllowedAt > timestamp && latestReleaseCache.release) {
+    return latestReleaseCache.release;
+  }
+
+  // Abort the request after 30 seconds
+  const controller = new AbortController();
+  const timeout = setTimeout(() => controller.abort(), 30_000);
+
+  /* Fetch the latest release from the Github API
+   * https://developer.github.com/v3/repos/releases/#get-the-latest-release
+   */
+
+  const response = await fetch(LATEST_RELEASE_API_URL, {
```

Comment:
This won't work, while this promise is being awaited, other requests may come in, and the `nextFetchAllowedAt` has not updated yet. Resulting in multiple fetches. Instead, you need to cache the promise. For example:

```tsx
async function fetchRelease() {
  const response = await fetch(LATEST_RELEASE_API_URL, {
    // @ts-expect-error See https://github.com/microsoft/TypeScript/issues/48003
    signal: AbortSignal.timeout(30000),
  });

  if (response.ok) {
    const { tag_name: tag, html_url: url } = await response.json();
    return { tag, url };
  }

  throw new Error(`Failed to fetch latest release from Github API: ${response.status}`);
}

let nextFetchAllowedAt = 0;
let cached;

export async function getLatestToolpadRelease(): Promise<GithubRelease> {
  const now = Date.now();
  if (nextFetchAllowedAt < now || !cached) {
    cached = Promise.resolve().then(async () => {
      try {
        return await fetchRelease();
      } catch (error) {
        // retry in 30 seconds
        nextFetchAllowedAt = Date.now() + 1000 * 30;
        throw error;
      }
    });

    nextFetchAllowedAt = now + 1000 * 60 * 10;
  }

  return cached;
}
```